File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bridge/base/FlutterObjectFactory.kt
package com.cleveradssolutions.plugin.flutter.bridge.base

import com.cleveradssolutions.plugin.flutter.util.getArgAndReturn
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

abstract class FlutterObjectFactory<T>(
    binding: FlutterPlugin.FlutterPluginBinding,
    channelName: String
) : MethodHandler(binding, channelName) {

    private val map: MutableMap<String, T> = HashMap()

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<String>("id", result) { id ->
            when (call.method) {
                "initObject" -> map[id] = initInstance(id)
                "disposeObject" -> map.remove(id)
                else -> super.onMethodCall(call, result)
            }
        }
    }

    abstract fun initInstance(id: String): T

    operator fun get(id: String): T? = map[id]

}

File: ios/Classes/Bridge/Base/FlutterObjectFactory.swift
//
//  MethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import Flutter

class FlutterObjectFactory<T>: MethodHandler {
    private var map: [String: T] = [:]

    override func onMethodCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("id", result) { id in
            switch call.method {
            case "initObject": map[id] = initInstance(id: id)
            case "disposeObject": map.removeValue(forKey: id)
            default: super.onMethodCall(call, result)
            }
        }
    }

    open func initInstance(id: String) -> T {
        fatalError("FlutterObjectFactory<T>.initInstance(id:) must be overridden in a subclass!")
    }

    subscript(id: String) -> T? {
        return map[id]
    }
}
